amethyst_core/src/transform/systems.rs

```diff
@@ -31,87 +29,92 @@ impl TransformSystem {
 
-    fn run(&mut self, (entities, locals, children, mut init, mut globals): Self::SystemData) {
         // Clear dirty flags on `Transform` storage, before updates go in
         (&mut globals).open().1.clear_flags();
 
-        #[cfg(feature = "profiler")]
-        profile_scope!("transform_system");
-        // Checks for entities with a local transform and parent, but no
-        // `Init` component.
-        for (entity, _, child, _) in (&*entities, &locals, &children, !&init).join() {
-            self.indices.insert(entity, self.sorted.len());
-            self.sorted.push((entity, child.parent()));
-            self.new.push(entity);
-        }
-
-        // Deletes entities whose parents aren't alive.
-        for &(entity, _) in &self.sorted {
-            if let Some(child) = children.get(entity) {
-                if !entities.is_alive(child.parent()) || self.dead.contains(&child.parent()) {
-                    let _ = entities.delete(entity);
-                    self.dead.insert(entity);
+        {
+            for (entity, parent) in (&*entities, parents.open().1).join() {
+                if parent.entity == entity {
+                    self.remove_parent.push(entity);   
                 }
             }
+
+            for entity in self.remove_parent.iter() {
+                eprintln!("Entity was its own parent: {:?}", entity);
+                parents.remove(*entity);
+            }
+
+            self.remove_parent.clear();
         }
 
-        // Adds an `Init` component to the entity.
-        for entity in self.new.drain(..) {
-            init.insert(entity, Init);
+        {
+            // Checks for entities with a modified local transform or a modified parent, but isn't initialized yet.
+            let filter = locals.open().0 & parents.open().0 & !&self.init; // has a local, parent, and isn't initialized.
+            for (entity, _) in (&*entities, &filter).join() {
+                self.indices.insert(entity, self.sorted.len());
+                self.sorted.push(entity);
+                self.frame_init.add(entity.id());
+            }
         }
 
         {
+            let locals_flagged = locals.open().1;
+
             // Compute transforms without parents.
-            for (ent, local, (mut global_entry, global_restrict), _) in
-                (&*entities, &locals, &mut globals.restrict(), !&children).join()
-            {
-                if local.is_dirty() {
-                    self.dirty.insert(ent);
-                    let global = global_restrict.get_mut_unchecked(&mut global_entry);
-                    global.0 = local.matrix();
-                    local.flag(false);
-                }
+            for (_entity, local, global, _) in (&*entities, locals_flagged, &mut globals, !&parents).join() {
+                global.0 = local.matrix();
+                debug_assert!(global.0 == global.0, format!("Entity {:?} had NaN transform.", _entity));
             }
```

Comment:
Should be `is_finite`

Reply:
I don't think `[[f32; 4]; 4]` has methods implemented by default? `f32` has `is_finite`, but not this.

Reply:
Ah right. I wonder if this is valid then? I'm not sure if the total equality check catches everything `is_finite` does

Reply:
It doesn't check for Infinity, but I'm not sure if it matters if we allow that or not.

Reply:
If that's the case this check doesn't make too much sense.